### PR TITLE
Add EnsureCapacity to ArrayPoolExtensions

### LINKED ISSUE #191: Api Proposal: ArrayPool extension "EnsureCapacity"
### Overview

When working with ReadOnlySequences or converting strings to binary data (or vice versa) I've found it useful to use a pooled array as a temporary buffer when needed.

The buffer should be rented only when needed and sliced if it's too big, which is a different use case compared to the existing `Resize`-extension.
Doing null/length checks manually is not complex, but possibly a bit error prone and verbose.

I'm not overly attached to the `EnsureCapacity`-name, but it's the best I've come up with. The functionality is also loosely related to EnsureCapacity on List and Dictionary.

### API breakdown

Ensures that when the method returns, `array` is not null and is at least `capacity` in length. In case the array needs to be "resized", contents are not preserved.

Error handling for negative capacity can be left to the pool.

```csharp
namespace CommunityToolkit.HighPerformance;

public static class ArrayPoolExtensions
{
    public static void EnsureCapacity<T>(
        this ArrayPool<T> pool,
        [NotNull] ref T[]? array,
        int capacity,
        bool clearArray = false)
    {
        if (array is null)
        {
            array = pool.Rent(capacity);
        }
        else if (array.Length < capacity)
        {
            var newArray = pool.Rent(capacity);
            pool.Return(array, clearArray);
            array = newArray;
        }
    }
}
```

### Usage example

```csharp
byte[]? buffer = null;

try
{
    foreach (ReadOnlySequence<byte> sequence in SomeReader())
    {
        if (sequence.IsSingleSegment)
        {
            ParseSpan(sequence.FirstSpan);
        }
        else
        {
            ArrayPool<byte>.Shared.EnsureCapacity(ref buffer, (int)sequence.Length);
            sequence.CopyTo(buffer);
            ParseSpan(buffer.AsSpan(0, (int)sequence.Length));
        }
    }
}
finally
{
    if (buffer is not null)
        ArrayPool<byte>.Shared.Return(buffer);
}
```

### Breaking change?

No

### Alternatives

The method is relatively simple, possibly rarely used, and can be copy-pasted wherever it's needed; simplest alternative is to just not implement it.

I've implemented it a couple of times in projects already referencing the HighPerformance-package. I can live with it if the extra api surface isn't worth the cost.

I'm interested in hearing everyone's thoughts on this. 

### Additional context

_No response_

### Help us help you

Yes, I'd like to be assigned to work on this item

### PR DESCRIPTION
**Pull request by ovska**
**Closes #191**

Implements an extension `EnsureCapacity` for `ArrayPool<T>` that ensures that an array is not null and has a minimum length. Documentation closely mirrors the existing `Resize`-method for consistency. The only difference to the original proposal is the range check for `capacity`, as requested.

~~Added `[DoesNotReturn]` "polyfill" for under NETSTANDARD2.1. This isn't critical and can be removed if deemed not to be worth it.~~

## PR Checklist

- [x] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [x] Based off latest main branch of toolkit
- [x] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [x] Tested code with current [supported SDKs](../#supported)
- [x] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [x] Every new API (including internal ones) has full XML docs
- [x] Code follows all style conventions

## Other information

Renamed weird/outdated test method names for existing `ArrayPoolExtensions.Resize`-tests so new tests fit in better.

### CODE REVIEW

File: CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs [hunk 1]
```diff
@@ -53,4 +53,41 @@ public static void Resize<T>(this ArrayPool<T> pool, [NotNull] ref T[]? array, i
 
         array = newArray;
     }
+
+    /// <summary>
+    /// Ensures that when the method returns <paramref name="array"/> is not null and is at least <paramref name="capacity"/> in length.
+    /// Contents of <paramref name="array"/> are not copied if a new array is rented.
+    /// </summary>
+    /// <typeparam name="T">Array item type</typeparam>
```
Comment (Sergio0694) [inline CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs:61]:
```suggestion
    /// <typeparam name="T">The type of items into the target array given as input.</typeparam>
```

File: CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs [hunk 2]
```diff
@@ -53,4 +53,41 @@ public static void Resize<T>(this ArrayPool<T> pool, [NotNull] ref T[]? array, i
...
+    /// <param name="pool">The target <see cref="ArrayPool{T}"/> instance used to rent and/or return the array.</param>
+    /// <param name="array">The rented <typeparamref name="T"/> array to ensure capacity for, or <see langword="null"/> to rent a new array.</param>
+    /// <param name="capacity">Minimum length of <paramref name="array"/> when the method returns.</param>
```
Comment (Sergio0694) [inline CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs:64]:
```suggestion
    /// <param name="capacity">The minimum length of <paramref name="array"/> when the method returns.</param>
```

File: CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs [hunk 3]
```diff
@@ -53,4 +53,41 @@ public static void Resize<T>(this ArrayPool<T> pool, [NotNull] ref T[]? array, i
...
+            ThrowNegativeCapacity();
+        }
+
+        if (array is null)
+        {
+            array = pool.Rent(capacity);
+        }
+        else if (array.Length < capacity)
+        {
+            // ensure rent succeeds before returning the original array
```
Comment (Sergio0694) [inline CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs:81]: Nit: comments should start in uppercase
```suggestion
            // Ensure rent succeeds before returning the original array
```

File: CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs [hunk 4]
```diff
@@ -53,4 +53,41 @@ public static void Resize<T>(this ArrayPool<T> pool, [NotNull] ref T[]? array, i
...
+            T[] newArray = pool.Rent(capacity);
+            pool.Return(array, clearArray);
+            array = newArray;
```
Comment (Sergio0694) [inline CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs:84]: Code style: add a blank line between declarations, invocations and assignments.
```suggestion
            T[] newArray = pool.Rent(capacity);

            pool.Return(array, clearArray);

            array = newArray;
```

File: CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs [hunk 5]
```diff
@@ -53,4 +53,41 @@ public static void Resize<T>(this ArrayPool<T> pool, [NotNull] ref T[]? array, i
...
+        }
+
+        [DoesNotReturn]
+        static void ThrowNegativeCapacity()
+        {
+            throw new ArgumentOutOfRangeException(nameof(capacity));
```
Comment (Sergio0694) [inline CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs:90]: Given this is a throw helper anyway, we might as well add a helpful error message as well:
```suggestion
            throw new ArgumentOutOfRangeException(nameof(capacity), "The array capacity must be a positive number.");
```

File: CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs [hunk 6]
```diff
@@ -53,4 +53,41 @@ public static void Resize<T>(this ArrayPool<T> pool, [NotNull] ref T[]? array, i
...
+        else if (array.Length < capacity)
+        {
+            // ensure rent succeeds before returning the original array
+            T[] newArray = pool.Rent(capacity);
+            pool.Return(array, clearArray);
+            array = newArray;
+        }
+
+        [DoesNotReturn]
+        static void ThrowNegativeCapacity()
```
Comment (Sergio0694) [inline CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs:88]: Nit: use a more expressive method name:

```suggestion
        static void ThrowArgumentOutOfRangeExceptionForNegativeCapacity()
```

File: tests/CommunityToolkit.HighPerformance.UnitTests/Extensions/Test_ArrayPoolExtensions.cs [hunk 7]
```diff
@@ -87,4 +87,62 @@ public void Test_ArrayExtensions_Clear()
...
+    [ExpectedException(typeof(ArgumentOutOfRangeException))]
+    public void Test_ArrayPoolExtensions_EnsureCapacity_InvalidSize()
+    {
+        int[]? array = null;
+
+        ArrayPool<int>.Shared.EnsureCapacity(ref array, -1);
+    }
+
+    [TestMethod]
+    public void Test_ArrayPoolExtensions_EnsureCapacity_NewArray()
```
Comment (Sergio0694) [inline tests/CommunityToolkit.HighPerformance.UnitTests/Extensions/Test_ArrayPoolExtensions.cs:101]: Could you also add a test with the input array being of exactly the same capacity as the new requested capacity?

File: CommunityToolkit.HighPerformance/Properties/Polyfills/DoesNotReturnAttribute.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if !NETSTANDARD2_1_OR_GREATER
+
+namespace System.Diagnostics.CodeAnalysis;
+
+/// <summary>Applied to a method that will never return under any circumstance.</summary>
+[AttributeUsage(AttributeTargets.Method, Inherited = false)]
+public sealed class DoesNotReturnAttribute : Attribute { }
```
Comment (Sergio0694) [inline CommunityToolkit.HighPerformance/Properties/Polyfills/DoesNotReturnAttribute.cs:11]: This cannot be public, at best it should be internal. But regardless, I don't think we should add this in this PR, given it's not really needed and all the other throw helpers in the paclage are not using this attribute either. If we really wanted to do this, it should be a separate follow up PR just doing that 🙂

**Review by Sergio0694 — CHANGES_REQUESTED** [review]
Looking good! Left some minor nits 😊

File: CommunityToolkit.HighPerformance/Properties/Polyfills/DoesNotReturnAttribute.cs (hunk 8 above)
Comment (ovska) [reply CommunityToolkit.HighPerformance/Properties/Polyfills/DoesNotReturnAttribute.cs]: Good point, I've removed the attribute.

**Comment by ovska** [comment]
Thanks for the review, I've fixed the nits!

File: tests/CommunityToolkit.HighPerformance.UnitTests/Extensions/Test_ArrayPoolExtensions.cs (hunk 7 above)
Comment (ovska) [reply tests/CommunityToolkit.HighPerformance.UnitTests/Extensions/Test_ArrayPoolExtensions.cs:112]: I've added a test for EnsureCapacity after Rent with the same size; not sure if it's worth it to test more precisely as the exact size of the rented array isn't carved in stone. Let me know if you think of something.

File: CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs [hunk 9]
```diff
@@ -53,4 +53,42 @@ public static void Resize<T>(this ArrayPool<T> pool, [NotNull] ref T[]? array, i
...
+        {
+            // Ensure rent succeeds before returning the original array to the pool
+            T[] newArray = pool.Rent(capacity);
+
+            pool.Return(array, clearArray);
+
+            array = newArray;
+        }
+
+        static void ThrowArgumentOutOfRangeExceptionForNegativeCapacity()
```
Comment (Sergio0694) [inline CommunityToolkit.HighPerformance/Extensions/ArrayPoolExtensions.cs:89]: Nit: due to Roslyn not optimizing this, using a local function here causes an unnecessary generic parameter being introduced here ([see here](https://github.com/dotnet/roslyn/issues/55800#issuecomment-1092965326)). Could you move this throw helper to a private static method in this class, outside of this method, and also add full XML docs on it while at it? You can see here as a reference:

https://github.com/CommunityToolkit/dotnet/blob/f3bd3d8b72615f1c1a5cd02876518ef01de518e5/CommunityToolkit.HighPerformance/Memory/Internals/ThrowHelper.cs#L74-L80

**Review by Sergio0694 — CHANGES_REQUESTED** [review]
Looks great! Thank you for following the guide and force pushing as well 😄
Left one minor nit (not your fault, just Roslyn not being really helpful here).

**Review by Sergio0694 — APPROVED** [review]
LGTM, thanks! 🚀

**Pull request merged**